File: Specializations/Data Structures and Algorithms/Algorithmic Toolbox/Week 2/Assignment/Haskell/Lcm.hs
{-# LANGUAGE BangPatterns #-}
import System.IO
import Control.Monad

main :: IO ()
main = do
    hSetBuffering stdout NoBuffering

    input_line <- getLine
    let split :: Eq a => a -> [a] -> [[a]]
        split d [] = []
        split d s = x : split d (drop 1 y) where (x,y) = span (/= d) s

    let splitted = split ' ' input_line
    let numbers = map (read::String->Int) splitted

    let lcm a b = lcmIter a b 1
            where
                lcmIter !a !b !l | l > a * b                    = a * b
                                 | mod l a == 0 && mod l b == 0 = l
                                 | otherwise                    = lcmIter a b (l + 1)

    let res = lcm (numbers!!0) (numbers!!1)

    
    putStrLn $ show res

    return ()

File: Specializations/Data Structures and Algorithms/Algorithmic Toolbox/Week 2/Assignment/Scala/src/main/scala/Lcm.scala
object Lcm {
  def lcm(a: Long, b: Long): Long = {
    def lcmIter(a: Long, b: Long, l: Long): Long = {
      if (l > a * b) a * b
      else if (l % a == 0 && l % b == 0) l
      else lcmIter(a, b, l + 1)
    }
    lcmIter(a, b, 1)
  }

  def main(args: Array[String]): Unit = {
    val input = scala.io.StdIn.readLine()
    val tokens = input.split(" ").map(_.toInt)
    println(lcm(tokens(0), tokens(1)))
  }
}
